tests/UNITTESTS/LinkLoss/test_LinkLossService.cpp

```diff
@@ -0,0 +1,414 @@
+}
+
+TEST_F(TestLinkLossServiceEvents, disconnection_mild_alert)
+{
+    // Set the alert level to "MILD ALERT"
+    link_loss_service->set_alert_level(LinkLossService::AlertLevel::MILD_ALERT);
+
+    // Simulate a clean connection event
+    simulate_connection_event(BLE_ERROR_NONE);
+
+    // The device should start alerting with an alert level of "MILD ALERT"
+    EXPECT_CALL(event_handler_mock, on_alert_requested(LinkLossService::AlertLevel::MILD_ALERT))
+         .Times(1);
+
+    // Simulate a disconnection event due to connection timeout
+    simulate_disconnection_event(disconnection_reason_t::CONNECTION_TIMEOUT);
+
+    // Dispatch events for 59999 ms
+    event_queue.dispatch(59999);
+
+    // The alert should end in 1 ms due to the alert timeout
+    EXPECT_CALL(event_handler_mock, on_alert_end())
+         .Times(1);
+
+    // Dispatch events for a further 1 ms
+    event_queue.dispatch(1);
+}
+
+TEST_F(TestLinkLossServiceEvents, disconnection_high_alert)
+{
+    // Set the alert level to "HIGH ALERT"
+    link_loss_service->set_alert_level(LinkLossService::AlertLevel::HIGH_ALERT);
+
+    // Simulate a clean connection event
+    simulate_connection_event(BLE_ERROR_NONE);
+
+    // The device should start alerting with an alert level of "HIGH ALERT"
+    EXPECT_CALL(event_handler_mock, on_alert_requested(LinkLossService::AlertLevel::HIGH_ALERT));
+
+    // Simulate a disconnection event due to connection timeout
+    simulate_disconnection_event(disconnection_reason_t::CONNECTION_TIMEOUT);
+
+    // Dispatch events for 59999 ms
+    event_queue.dispatch(59999);
+
+    // The alert should end in 1 ms due to the alert timeout
+    EXPECT_CALL(event_handler_mock, on_alert_end())
+         .Times(1);
+
+    // Dispatch events for a further 1 ms
+    event_queue.dispatch(1);
+}
+
+TEST_F(TestLinkLossServiceEvents, disconnection_reconnection)
+{
+    // Set the alert level to "HIGH ALERT"
+    link_loss_service->set_alert_level(LinkLossService::AlertLevel::HIGH_ALERT);
+
+    // Simulate a clean connection event
+    simulate_connection_event(BLE_ERROR_NONE);
+
+    // The device should start alerting with an alert level of "HIGH ALERT"
+    EXPECT_CALL(event_handler_mock, on_alert_requested(LinkLossService::AlertLevel::HIGH_ALERT));
+
+    // Simulate a disconnection event due to connection timeout
+    simulate_disconnection_event(disconnection_reason_t::CONNECTION_TIMEOUT);
+
+    // Dispatch events for 59999 ms
+    event_queue.dispatch(59999);
+
+    // Get the number of events in the queue before reconnection
+    size_t initial_queue_size = event_queue.size();
+
+    // The alert should end due to reconnection
+    EXPECT_CALL(event_handler_mock, on_alert_end())
+         .Times(1);
+
+    // Simulate a clean connection event
+    simulate_connection_event(BLE_ERROR_NONE);
+
+    // The reconnection should have cancelled the pending timeout
+    ASSERT_EQ(event_queue.size(), initial_queue_size - 1);
```

Comment:
But the re-connection cancels the pending timeout. If I move that check before the re-connection, it will fail. 

Adding an expectation `on_alert_end()` not being called before re-connection is not an issue.

Reply:
It should be fine, that extra check is implicit anyway.